pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -370,12 +472,79 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 	}
 	var err error
 	for k, v := range configs {
-		resources, err = addConfigMapItem(master, getConfigName(master, k), labels, v, resources)
+		resources, err = addConfigMapItem(master, getResourceName(master, k), labels, v, resources)
 		if err != nil {
 			return nil, err
 		}
 	}
 
+	// User has reset the CR after failure. Remove upgrade failed status.
+	if master.Spec.Image == master.Status.ImageToUse && master.Status.Meta.IsConditionTrue(upgradeFailed) {
+		master.Status.ClearCondition(upgradeFailed, upgradeResetMessage, upgradeResetMessage)
+	}
+
+	if !imageToUseNeedsUpdate(master) {
+		// Set ImageToUse if upgrade has not failed
+		if !master.Status.Meta.IsConditionTrue(upgradeFailed) {
+			master.Status.ImageToUse = master.Spec.Image
```

Comment:
This logic seems contradicting, as it is read as "if the image to use doesn't need update (line 486), update the image to use (line 489)".